### PR TITLE
enable stepwise execution of query pipelines

### PR DESCRIPTION
**Pull request by logan-markewich**
QueryPipelines are extremely stateful. If another module tries to drive the execution of a pipeline, they will have a hard time doing with without using private methods and reading a lot of source code

Here's a sample of what it looks like

```python
run_state = p.get_run_state(movie_name="The Departed")

next_module_keys = p.get_next_module_keys(run_state)

while True:
    for module_key in next_module_keys:
        # get the module and input
        module = run_state.module_dict[module_key]
        module_input = run_state.all_module_inputs[module_key]

        # run the module
        output_dict = module.run_component(**module_input)

        # process the output
        p.process_component_output(
            output_dict,
            module_key,
            run_state,
        )

    # get the next module keys
    next_module_keys = p.get_next_module_keys(
        run_state,
    )

    # if no more modules to run, break
    if not next_module_keys:
        run_state.result_outputs[module_key] = output_dict
        break

# the final result is at `module_key`
# it is a dict of 'output' -> ChatResponse object in this case
print(run_state.result_outputs[module_key]["output"].message.content)
```

### CODE REVIEW

File: llama-index-core/llama_index/core/query_pipeline/query.py [hunk 1]
```diff
@@ -609,6 +610,7 @@ def _get_single_result_output(
 
         """
         if len(result_outputs) != 1:
+            print(result_outputs)
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/query_pipeline/query.py:613]: nit: prints

File: llama-index-core/llama_index/core/query_pipeline/query.py [hunk 2]
```diff
@@ -744,146 +746,106 @@ def _validate_inputs(self, module_input_dict: Dict[str, Any]) -> None:
                 f"Input keys: {module_input_dict.keys()}\n"
             )
 
-    def _process_component_output(
+    def process_component_output(
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/query_pipeline/query.py:749]: is this user-facing? when would they use this?

**Review by jerryjliu — COMMENTED** [review]
out of curiosity what's the user-facing UX? (was this meant to be a user-facing PR? )

File: llama-index-core/llama_index/core/query_pipeline/query.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/query_pipeline/query.py:774]: I have greatly improved the UX here now, and added a notebook example

**Review by jerryjliu — COMMENTED** [review]
i know you're exposing these as public methods to make it easier to plug into from an outer service

from a user perspective though the full while loop might be a bit too low level for users to use. I wonder if besides the current methods, we can also define a `run_step` method which does something like 

```python
run_state = p.get_run_state(movie_name="The Departed")
output_dict = p.run_step(run_state)
```

where run_step just wraps something like this
```
for module_key in next_module_keys:
        # get the module and input
        module = run_state.module_dict[module_key]
        module_input = run_state.all_module_inputs[module_key]

        # run the module
        output_dict = module.run_component(**module_input)

        # process the output
        p.process_component_output(
            output_dict,
            module_key,
            run_state,
        )
```

**Review by jerryjliu — APPROVED** [review]
(thinking about it though the above suggestion could always be a future PR, so approving to unblock)

**Comment by logan-markewich** [comment]
Good point. We can improve before advertising this api

**Pull request merged**